File: backend/core/erc8183_negotiate.py
"""
erc8183_negotiate.py

Real, server-side proxy for the ERC-8183 `negotiate` A2A skill — the missing
step that caused a real, confirmed incident (2026-08-21/22): job #56636 was
funded through the marketplace's generic hire flow (a fixed plain-text
description, no negotiate step) and was then PERMANENTLY rejected by the
explainer agent's own `notify_funded` handler with "no signed quote anchored
in job description". Traced through the real SDK verification logic
(bnbagent_studio_core.erc8183.verify.verify_signed_job): a strict ERC-8183
seller requires the on-chain job description to be a Schema-v1 JobDescription
carrying a `negotiation_hash` + `provider_sig` this exact provider signed —
which only the `negotiate` skill produces. Confirmed live (2026-08-22)
against the explainer agent's real negotiate response, run through the SDK's
own `build_job_description` + `recover_quote_signer`: the reconstructed
description round-trips and recovers the correct signer address.

Why this lives server-side, not called directly from the browser: confirmed
live (2026-08-22) that the explainer agent's endpoint has NO CORS support —
a real OPTIONS preflight returned 405 with no Access-Control-Allow-Origin
header, so a direct browser fetch would be blocked outright by the browser
itself, for every marketplace agent, not just this one (no agent built for
this marketplace advertises CORS headers; that's a reasonable default for a
server meant to be called server-to-server, same reasoning documented in
core/agent_health.py for why health checks run backend-side too).

Real, honest scope: this only works for agents that (a) have a real
`service_endpoint` on record (see core/agent_health.py) and (b) actually
implement the A2A `negotiate` skill the way this SDK's seller template does.
Many marketplace agents will not — this proxy returns None cleanly in every
failure case (no endpoint, endpoint isn't a real A2A agent card, timeout,
negotiation rejected) so the caller (server.py's endpoint, then
useHireAgent.js) can fall back to the existing plain-description flow
exactly as before, rather than break hiring for agents that never supported
negotiate in the first place.

Real SSRF gap found and fixed (2026-08-27 security audit): unlike
deliverable_proxy.py (guarded from the start — see that module's own real
threat-model note), this proxy had NO host-safety validation at all. While
`service_endpoint` is resolved server-side (server.py never accepts a raw
URL from the client directly, only an owner_address), its real VALUE is
still attacker-influenced: whoever registers an agent on-chain sets it, so
a malicious registrant could point it at internal infrastructure (a cloud
metadata IP, an internal admin service) and get every real user who tries
to hire that "agent" to make this backend proxy a request there — the same
real threat deliverable_proxy.py's own docstring describes, just via
registration instead of a job's submit(). Worse here: `_jsonrpc_candidates`
below ALSO trusts a `url` field pulled out of whatever JSON the endpoint's
own `.well-known/agent-card.json` happens to return — a second,
attacker-controlled hop that could point anywhere even if the original
service_endpoint looked benign. Fixed by reusing deliverable_proxy.py's
already-proven `_is_safe_public_host` guard (DNS-resolve-time private/
loopback/link-local/reserved/multicast rejection) at BOTH hops: before the
agent-card fetch, and before trusting any `url` the card itself claims.

Real, multi-format retry (2026-08-27), from a real, live investigation
against 3 genuinely different agents: this used to send exactly one real
message shape — a `{"kind": "data", "data": skill_data}` part — and give
up on anything else. Real, confirmed live findings that made this
insufficient: OnyxOracle's real endpoint (app.singularry.org/api/a2a) is a
genuinely working A2A server that rejects a `data` part outright
("Only text parts are accepted by this endpoint") but returns a real
HTTP 200 for the exact same skill_data wrapped in a `{"kind": "text",
"text": json.dumps(skill_data)}` part instead; AIDA's registered endpoint
doesn't implement A2A JSON-RPC at all (a real 405 on POST, not a format
problem); mandaterebalance-agent requires an OAuth2 Bearer token only its
own operator can issue (a real 401, also not a format problem). `data` and
`text` are the two real, spec-defined A2A Part kinds relevant to a skill
call (the third, `file`, doesn't apply here) — not a guess, and not
extended further than what's actually been observed live. Now tries both
shapes, in order, against each real candidate URL, and logs which one (if
any) got a genuine, structured acknowledgment — never fabricating success
for a reply that merely returned 200 without actually confirming it
understood the request (see _extract_structured_reply below for exactly
what counts as "genuine").
"""

from __future__ import annotations

import json
import uuid

import httpx

from core.deliverable_proxy import _is_safe_public_host

_TIMEOUT = 15.0

# Real, spec-defined A2A message Part "kind" values relevant to a skill
# call — see module docstring for why only these two, and why not more.
_PART_SHAPES = ("data", "text")


def _is_safe_url(url: str) -> bool:
    """Real, shared guard for any URL this module is about to fetch —
    scheme + resolved-host safety, same standard as deliverable_proxy.py."""
    from urllib.parse import urlparse
    parsed = urlparse(url)
    if parsed.scheme not in ("http", "https") or not parsed.hostname:
        return False
    return _is_safe_public_host(parsed.hostname)


async def _jsonrpc_candidates(service_endpoint: str, client: httpx.AsyncClient) -> list[str]:
    """Real candidate JSON-RPC URLs for `service_endpoint`, ordered by
    DEMONSTRATED reliability across the real, live agents this has
    actually been tested against — not by protocol purity.

    Real gap found and fixed (2026-08-27): this used to NEVER try
    `service_endpoint` itself as a direct POST target — only a same-origin
    root guess (tuned to our own explainer agent's own specific bug, see
    below) and whatever URL a fetched agent-card happened to claim. Real,
    live, confirmed miss: OnyxOracle's actual registered service_endpoint
    (app.singularry.org/api/a2a) IS the real, correct, directly-POSTable
    JSON-RPC target — a real, live POST there succeeds — but it was never
    even attempted, because it isn't the same-origin root and its own
    agent-card fetch didn't hand back a matching `url`. The standard A2A
    expectation is that the registered endpoint just works directly; our
    own agent's same-origin-root quirk (below) is real but was one agent's
    specific bug, not the general case, now confirmed by a second real
    agent that behaves the standard way instead. `service_endpoint` itself
    is now tried FIRST for that reason.

    Real, confirmed finding (2026-08-22), kept as a fallback: the agent
    card's own advertised `url` field is NOT trustworthy — live-checked
    against our own explainer agent and it reports `http://localhost:10000/`
    (a local-dev placeholder) even after correctly setting the documented
    AGENTCORE_RUNTIME_URL env var and restarting; this looks like a real
    bug in the installed `bedrock_agentcore` package's URL-override path,
    not a configuration problem on our side. That project's own real A2A
    convention (`bedrock_agentcore.runtime.a2a`, confirmed by reading its
    source) serves JSON-RPC at the SAME origin's root path "/" regardless
    of what the card claims — kept as the second real candidate."""
    from urllib.parse import urlparse

    parsed = urlparse(service_endpoint)
    candidates: list[str] = []
    if _is_safe_url(service_endpoint):
        candidates.append(service_endpoint)

    same_origin = f"{parsed.scheme}://{parsed.netloc}/" if parsed.scheme and parsed.netloc else None
    if same_origin and same_origin not in candidates and _is_safe_url(same_origin):
        candidates.append(same_origin)

    card_urls = [service_endpoint]
    if parsed.scheme and parsed.netloc:
        card_urls.append(f"{parsed.scheme}://{parsed.netloc}/.well-known/agent-card.json")
    for url in card_urls:
        # Real SSRF guard, first hop — see module docstring.
        if not _is_safe_url(url):
            continue
        try:
            resp = await client.get(url, timeout=_TIMEOUT, follow_redirects=False)
            resp.raise_for_status()
            card = resp.json()
        except Exception:
            continue
        rpc_url = card.get("url")
        # Real SSRF guard, second hop — the card's own claimed `url` is
        # JUST as attacker-controlled as service_endpoint itself (see
        # module docstring), so it gets the exact same real check before
        # ever being added as something we'll later POST to.
        if isinstance(rpc_url, str) and rpc_url and rpc_url not in candidates and _is_safe_url(rpc_url):
            candidates.append(rpc_url)
        break  # first real card fetch that succeeds is enough

    return candidates


def _build_payload(skill_data: dict, shape: str) -> dict:
    """Real, spec-defined A2A message envelope, varying only the Part
    `kind` — see module docstring for why exactly these two shapes."""
    if shape == "data":
        part = {"kind": "data", "data": skill_data}
    elif shape == "text":
        # Real, live-confirmed alternative (OnyxOracle): the same real
        # skill_data, JSON-stringified into a plain text part instead of a
        # structured data part.
        part = {"kind": "text", "text": json.dumps(skill_data)}
    else:
        raise ValueError(f"unknown A2A part shape: {shape}")
    return {
        "jsonrpc": "2.0",
        "id": "1",
        "method": "message/send",
        "params": {
            "message": {
                "kind": "message",
                "messageId": str(uuid.uuid4()),
                "role": "user",
                "parts": [part],
            }
        },
    }


def _extract_structured_reply(body: dict) -> dict | None:
    """Real, honest extraction — deliberately conservative. Returns
    something only when the reply is GENUINELY structured: a real
    `data`-kind part (used as-is, the strict/original contract), or a real
    `text`-kind part whose own content happens to be valid, parseable JSON
    (some agents that insist on text parts still reply with real JSON
    inside one). A real, live, generic-sounding text reply that ISN'T
    valid JSON (confirmed live: OnyxOracle's own identity/capability blurb,
    sent back regardless of what we asked) returns None here — the whole
    point of this function is to never treat "the HTTP call succeeded" as
    "the agent genuinely acknowledged this specific request", which would
    be fabricating success where none occurred."""
    try:
        parts = body["result"]["parts"]
    except (KeyError, TypeError):
        return None
    for p in parts:
        if p.get("kind") == "data" and isinstance(p.get("data"), dict):
            return p["data"]
    for p in parts:
        if p.get("kind") == "text" and isinstance(p.get("text"), str):
            try:
                parsed = json.loads(p["text"])
            except (json.JSONDecodeError, TypeError):
                continue
            if isinstance(parsed, dict):
                return parsed
    return None


async def _call_skill(service_endpoint: str, skill_data: dict) -> tuple[dict, str] | None:
    """Shared A2A `message/send` plumbing for both `negotiate` and
    `notify_funded`: resolve the real JSON-RPC candidates, and for EACH
    one, try every real known message shape (_PART_SHAPES) in order —
    real, multi-format retry (2026-08-27), not giving up after one fixed
    shape. Returns (data, shape) on the first genuine, structured reply
    (see _extract_structured_reply), logging exactly which real
    candidate+shape combination worked — or None if every real combination
    was tried and none produced a genuine acknowledgment (no endpoint, not
    a real A2A agent, timeout, auth rejection, or a real reply we can't
    honestly parse as structured). Callers apply their own skill-specific
    acceptance check on the result."""
    if not service_endpoint:
        return None
    async with httpx.AsyncClient(timeout=_TIMEOUT) as client:
        candidates = await _jsonrpc_candidates(service_endpoint, client)
        if not candidates:
            return None

        for rpc_url in candidates:
            # Real, defense-in-depth SSRF guard — candidates are already
            # filtered in _jsonrpc_candidates above, but re-checking here
            # too means this loop is safe even if a future change ever
            # adds a candidate through a different path.
            if not _is_safe_url(rpc_url):
                print(f"[erc8183_negotiate] candidate {rpc_url} failed the real host-safety check, skipping")
                continue

            for shape in _PART_SHAPES:
                payload = _build_payload(skill_data, shape)
                try:
                    resp = await client.post(rpc_url, json=payload, timeout=_TIMEOUT, follow_redirects=False)
                    resp.raise_for_status()
                    body = resp.json()
                except Exception as e:
                    print(f"[erc8183_negotiate] {rpc_url} ({shape} part) failed, trying next real format: {e}")
                    continue

                if "error" in body:
                    print(f"[erc8183_negotiate] {rpc_url} ({shape} part): real JSON-RPC error, trying next real format: {body['error']}")
                    continue

                data = _extract_structured_reply(body)
                if data is not None:
                    print(f"[erc8183_negotiate] {rpc_url} ({shape} part): genuine, structured reply received — using this real format")
                    return data, shape

                print(f"[erc8183_negotiate] {rpc_url} ({shape} part): real HTTP 200 but no genuine structured reply "
                      f"(not treated as acceptance): {json.dumps(body)[:300]}")

        return None


async def negotiate(
    service_endpoint: str, task_description: str, terms: dict
) -> dict | None:
    """Real A2A `negotiate` call against a seller agent. Returns the raw
    negotiation-result dict (the same shape `NegotiationResult.to_dict()`
    produces — request/response/negotiation_hash/provider_sig/chain_id/
    verifying_contract) on a real accepted quote, or None on ANY failure
    (no endpoint, not a real A2A agent, timeout, or a genuine rejection) —
    the caller's job is to fall back cleanly, not to distinguish why."""
    result = await _call_skill(service_endpoint, {
        "skill": "negotiate",
        "task_description": task_description,
        "terms": terms,
    })
    if result is None:
        return None
    data, shape = result

    response = data.get("response") or {}
    if not response.get("accepted"):
        # A real, genuine rejection (e.g. malformed terms) — not a
        # transport failure. Still None: the caller falls back the same
        # way either way, and the real reason is already logged for us.
        print(f"[erc8183_negotiate] negotiation not accepted (real {shape}-part reply): {response}")
        return None
    print(f"[erc8183_negotiate] negotiation accepted via a real {shape}-part message")
    return data


async def notify_funded(service_endpoint: str, job_id: int, authorization: dict | None = None) -> dict | None:
    """Real A2A `notify_funded` push: "I funded job X — please deliver."

    Real, confirmed gap (2026-08-24): this marketplace's own hire flow
    (useHireAgent.js) creates + funds the on-chain job but never sent this
    notification — confirmed live against job #56646, which sat funded with
    zero delivery activity in the seller's own logs until this call was sent
    manually. A strict ERC-8183 seller (like our own explainer agent) has no
    other trigger to start work: its background "sweep" for missed funded
    jobs only runs as a side effect of ANOTHER buyer's notify_funded landing
    first (see explainer-agent/seller_core.py's own docstring) — with no
    other buyer ever notifying, a job funded through this marketplace could
    sit forever.

    Best-effort by design, same as `negotiate`: returns None only on a
    TRANSPORT failure (no endpoint, unreachable, malformed reply) — the
    caller must NEVER treat that None as the hire itself failing, the job
    is already funded on-chain regardless. Unlike `negotiate`, a REAL reply
    from the agent (accepted OR rejected) is returned as-is rather than
    collapsed to None — a rejection reason (e.g. "authorization_required",
    "caller_not_job_client") is real, useful signal for the caller/frontend
    to surface, not something to swallow.

    `authorization` (optional): a real, EIP-712-signed envelope for sellers
    that require one — real, confirmed example (2026-08-24): the live
    `stockanalyst-agent` (bnb-chain/stockanalyst-agent-demo pattern)
    unconditionally rejects notify_funded with "authorization_required"
    unless this exact dict is present: {"context": <json string>,
    "expires_at": <int>, "nonce": "0x"+64 hex, "signature": 130 hex (with
    or without "0x")} — verified against that project's real
    notify_security.py (server-side EIP-712 recovery + expected-client
    check). The signature must be produced client-side by the JOB'S OWN
    CLIENT WALLET (only the frontend, with the connected wallet, can sign
    it — this function only forwards an already-built envelope, never
    builds or signs one itself)."""
    skill_data = {"skill": "notify_funded", "job_id": job_id}
    if authorization is not None:
        skill_data["authorization"] = authorization
    result = await _call_skill(service_endpoint, skill_data)
    if result is None:
        return None
    data, shape = result
    if data.get("status") != "accepted":
        print(f"[erc8183_negotiate] notify_funded not accepted (real {shape}-part reply): {data}")
    else:
        print(f"[erc8183_negotiate] notify_funded accepted via a real {shape}-part message")
    return data

File: frontend/src/erc8183Negotiate.js
// erc8183Negotiate.js
//
// Real client-side counterpart to backend/core/erc8183_negotiate.py.
//
// Real root cause this exists for (2026-08-21/22): job #56636 was funded
// through the old hire flow — a fixed plain-text description, no negotiate
// step — and was then PERMANENTLY rejected by the explainer agent's own
// notify_funded handler: "no signed quote anchored in job description".
// Traced through the real SDK verification logic
// (bnbagent_studio_core.erc8183.verify.verify_signed_job): a strict
// ERC-8183 seller requires the on-chain description to be a Schema-v1
// JobDescription carrying a negotiation_hash + provider_sig THIS provider
// signed — which only the real `negotiate` skill produces. This isn't
// specific to that one job; ANY hire against a properly strict seller
// through the old flow was dead on arrival.
//
// negotiateJob() below calls our OWN backend (never the agent directly —
// confirmed live that the agent's endpoint has no CORS support at all, a
// real OPTIONS preflight returned 405 with no Access-Control-Allow-Origin,
// so a browser fetch would be blocked outright). buildJobDescription() is a
// field-for-field port of the SDK's own
// bnbagent.erc8183.negotiation.build_job_description — verified
// byte-for-byte compatible against a real, live negotiation result
// (2026-08-22): the description this produces re-parses and its
// negotiation_hash correctly recovers the agent's real signer address via
// the same recover_quote_signer logic the seller uses to verify it. No
// keccak/checksum library needed here — `verifying_contract` is relayed
// verbatim from the negotiation result, which the agent already returns
// correctly EIP-55 checksummed (checksumming is idempotent, so re-deriving
// it client-side would be redundant, not more correct).

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';

/** Real call to our backend's negotiate proxy. Returns the raw negotiation
 * result on a real accepted quote, or null — NEVER throws. The caller
 * (useHireAgent.js) falls back to the plain-description flow on null,
 * exactly the same way whether the agent doesn't support negotiate, isn't
 * reachable, or genuinely rejected the terms — that distinction doesn't
 * change what the buyer should do next.
 *
 * `agentId` (optional, real, added 2026-08-28): real, confirmed bug fix —
 * without it, the backend used to resolve "the agent" by owner_address
 * ALONE, which is genuinely ambiguous whenever one real owner operates
 * more than one registered agent (confirmed live: 1,457 real owners
 * currently do) — a real hire could silently negotiate against a
 * completely different, wrong agent that just happens to share an owner.
 * Always pass it when the caller has it (every real caller here does). */
export async function negotiateJob(ownerAddress, agentId, taskDescription, terms) {
  try {
    const res = await fetch(`${API_BASE_URL}/api/agents/negotiate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ owner_address: ownerAddress, agent_id: agentId, task_description: taskDescription, terms }),
    });
    if (!res.ok) return null;
    const body = await res.json();
    return body.available ? body.negotiation_result : null;
  } catch (e) {
    return null;
  }
}

/** Real call to our backend's notify_funded proxy — tells a strict ERC-8183
 * seller "I funded job X, please deliver" right after the real on-chain
 * `fund` tx confirms. Real, confirmed gap fixed 2026-08-24: without this,
 * a job funded through this marketplace had no trigger to ever get
 * delivered (see backend/server.py's /api/agents/notify-funded for the
 * full trace). Best-effort, NEVER throws and NEVER blocks the hire — the
 * job is already funded on-chain by the time this is called, so a failure
 * here means "delivery may be slower", never "the hire failed".
 *
 * `authorization` (optional): pass the real object buildNotifyAuthorization
 * (below) returns, for sellers that unconditionally require one (real,
 * confirmed example: the live stockanalyst-agent). Returns
 * {notified, reason?} rather than a bare boolean — the real rejection
 * reason (e.g. "invalid_authorization", "caller_not_job_client") is useful
 * signal, not something to swallow.
 *
 * `agentId` (optional, real, added 2026-08-28): same real fix as
 * negotiateJob above — without it, the backend's owner-only lookup could
 * notify a completely different, wrong same-owner agent, leaving the
 * real, correct one never told its job was funded. */
export async function notifyFunded(ownerAddress, agentId, jobId, authorization = null) {
  try {
    const body = { owner_address: ownerAddress, agent_id: agentId, job_id: Number(jobId) };
    if (authorization) body.authorization = authorization;
    const res = await fetch(`${API_BASE_URL}/api/agents/notify-funded`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });
    if (!res.ok) return { notified: false, reason: `backend returned HTTP ${res.status}` };
    const parsed = await res.json();
    return { notified: parsed.notified === true, reason: parsed.reason };
  } catch (e) {
    return { notified: false, reason: e.message || String(e) };
  }
}

/** Real EIP-712 "notify_funded" authorization envelope for sellers that
 * unconditionally require one — real, confirmed example (2026-08-24): the
 * live stockanalyst-agent (bnb-chain/stockanalyst-agent-demo pattern)
 * rejects a named notify_funded call with "authorization_required" unless
 * this exact envelope is attached. Verified byte-for-byte against that
 * project's own real source — buyer-client/src/notify-auth.ts (the
 * envelope shape) and stockanalyst/app/agent/notify_security.py (the
 * server-side EIP-712 recovery + expected-client check this must satisfy):
 * domain {name:"stockanalyst-notify-funded", version:"1", chainId,
 * verifyingContract}, type NotifyFunded{jobId:uint256, context:string,
 * expiresAt:uint64, nonce:bytes32}.
 *
 * `context` defaults to "{}" — every field the seller reads from it
 * (delivery gateway, portfolio, risk profile) is optional server-side, and
 * this marketplace has no real portfolio/risk data to honestly sign over.
 *
 * `signTypedDataAsync` is wagmi's `useSignTypedData()` return value — only
 * the connected wallet that actually created/funded the job can produce a
 * signature the seller's expected_client check accepts (it recovers the
 * signer and compares against the job's real on-chain `client` address). */
export async function buildNotifyAuthorization({ chainId, verifyingContract, jobId, signTypedDataAsync, context = '{}' }) {
  const nonceBytes = new Uint8Array(32);
  crypto.getRandomValues(nonceBytes);
  const nonce = `0x${Array.from(nonceBytes).map((b) => b.toString(16).padStart(2, '0')).join('')}`;
  const expiresAt = Math.floor(Date.now() / 1000) + 300; // 5 min — matches the reference implementation's own window

  const signature = await signTypedDataAsync({
    domain: { name: 'stockanalyst-notify-funded', version: '1', chainId, verifyingContract },
    types: {
      NotifyFunded: [
        { name: 'jobId', type: 'uint256' },
        { name: 'context', type: 'string' },
        { name: 'expiresAt', type: 'uint64' },
        { name: 'nonce', type: 'bytes32' },
      ],
    },
    primaryType: 'NotifyFunded',
    message: { jobId: BigInt(jobId), context, expiresAt: BigInt(expiresAt), nonce },
  });

  return { context, expires_at: expiresAt, nonce, signature };
}

/** Mirrors bnbagent.erc8183.negotiation._sanitize_for_claim exactly:
 * '[' -> '(', ']' -> ')' (prevents injection into the UMA claim's
 * [REQUEST]/[RESPONSE]/[VERIFY] section markers), strips ASCII control
 * chars except tab/newline. Must match byte-for-byte — this text is part
 * of the signed content. */
function sanitizeForClaim(s) {
  if (typeof s !== 'string') return String(s);
  const swapped = s.split('[').join('(').split(']').join(')');
  return Array.from(swapped)
    .filter((ch) => {
      const code = ch.codePointAt(0);
      return code >= 0x20 || ch === '\t' || ch === '\n';
    })
    .join('');
}

/** Real port of build_job_description(negotiation_result) — the exact
 * on-chain description string a strict ERC-8183 seller can verify. Throws
 * (caller should treat as "negotiate effectively failed, fall back") if
 * the result is missing required fields — mirrors the Python function's
 * own ValueError cases, never silently produces a description that would
 * fail verification. */
export function buildJobDescription(negotiationResult) {
  const response = negotiationResult.response || {};
  const request = negotiationResult.request || {};
  if (!response.accepted) {
    throw new Error('Cannot build description from a rejected negotiation');
  }
  const responseTerms = response.terms || {};
  const price = responseTerms.price || '';
  const currency = responseTerms.currency || '';
  if (!price) throw new Error('Negotiation response missing price');
  if (!currency) throw new Error('Negotiation response missing currency');

  // Quality fields only — price/currency are separate top-level fields on
  // the real schema, never nested in `terms` (matches job #56620's real,
  // confirmed-working on-chain description, and the Python builder).
  const terms = {
    deliverables: sanitizeForClaim(responseTerms.deliverables || ''),
    quality_standards: sanitizeForClaim(responseTerms.quality_standards || ''),
  };
  if (Array.isArray(responseTerms.success_criteria) && responseTerms.success_criteria.length) {
    terms.success_criteria = responseTerms.success_criteria.map(sanitizeForClaim);
  }

  const negotiatedAt = negotiationResult.negotiated_at ?? response.negotiated_at ?? Math.floor(Date.now() / 1000);
  const quoteExpiresAt = negotiationResult.quote_expires_at ?? response.quote_expires_at;

  const content = {
    version: 1,
    negotiated_at: negotiatedAt,
    task: sanitizeForClaim(request.task_description || ''),
    terms,
    price,
    currency,
  };
  if (quoteExpiresAt != null) content.quote_expires_at = quoteExpiresAt;
  if (negotiationResult.chain_id != null) content.chain_id = negotiationResult.chain_id;
  // Relayed verbatim — see module docstring on why no re-checksum is needed.
  if (negotiationResult.verifying_contract) content.verifying_contract = negotiationResult.verifying_contract;
  if (negotiationResult.negotiation_hash) content.negotiation_hash = negotiationResult.negotiation_hash;
  if (negotiationResult.provider_sig) content.provider_sig = negotiationResult.provider_sig;

  // Key order doesn't need to match Python's sort_keys=True — a strict
  // seller's own verification re-parses this JSON and re-serializes
  // canonically from the PARSED object before re-hashing (confirmed by
  // reading bnbagent_studio_core.erc8183.verify.recover_quote_signer), so
  // only the VALUES need to match, not this string's own key ordering.
  return JSON.stringify(content);
}

/** The real, agreed price as a BigInt raw-unit amount — what verify_signed_job
 * actually checks the funded budget against (must be >= this). */
export function negotiatedPriceRaw(negotiationResult) {
  const price = negotiationResult?.response?.terms?.price;
  if (!price) return null;
  return BigInt(price);
}
